src/main/tut/3.2-adjunctions.md

```diff
@@ -0,0 +1,130 @@
+```Haskell
+swap       :: (a,b) -> (b,a)
+swap (a,b) = (b,a)
+```
+```tut:silent
+def swap[A, B]: ((A, B)) => (B, A) = {
+  case (a, b) => (b, a)
+}
+```
+................
+```Haskell
+return :: d -> m d
+```
+```scala
+def `return`[D]: D => M[D]
+```
+................
+```Haskell
+extract :: w c -> c
+```
+```scala
+def extract[C]: W[C] => C
+```
+................
+```Haskell
+class (Functor f, Representable u) =>
+      Adjunction f u | f -> u, u -> f where
+  unit         :: a -> u (f a)
+  counit       :: f (u a) -> a
+```
+```scala
+abstract class Adjunction[F[_], U[_]](
+    implicit val F: Functor[F],
+    val U: Representable[U]
+){
+  def unit[A](a: => A): U[F[A]]
+
+  def counit[A](a: F[U[A]]): A
+}
+```
+................
+```Haskell
+class (Functor f, Representable u) =>
+      Adjunction f u | f -> u, u -> f where
+  leftAdjunct  :: (f a -> b) -> (a -> u b)
+  rightAdjunct :: (a -> u b) -> (f a -> b)
+```
+```scala
+abstract class Adjunction[F[_], U[_]](
+    implicit val F: Functor[F],
+    val U: Representable[U]
+){
+  /* changed the order of parameters
+     to help scala compiler infer types */
+  def leftAdjunct[A, B]
+      (a: => A)(f: F[A] => B): U[B]
```

Comment:
same question wrt byname

Reply:
Changing parameter order makes sense from a scala perspective, but it might make seeing the relationship harder if one doesn't also read the haskell code

Reply:
That's why I've added a comment.

The reasons to change the order in scala version:
1. We are still going to do it in scala for sometime
2. This example: ```rightAdjunct(a)(identity)```